File: Fdk2R/FdkRHost/FdkPositions.cs

using System;
using System.Collections.Generic;
using System.Linq;
using FdkMinimal;
using SoftFX.Extended;
using log4net;
using RHost;

namespace FdkRHost
{
	/// <summary>
	/// Description of FdkPositions.
	/// </summary>
	public class FdkPosition
	{
		static DataTrade Trade
		{
			get { return FdkHelper.Trade; }
		}

		public static string GetTradePositions()
		{
			try
			{
                Position[] tradeRecords = Trade.Cache.Positions;
				
				Log.InfoFormat("FdkPositions.GetPositionPositions()");
				
				var varName = FdkVars.RegisterVariable(tradeRecords, "positions");
				return varName;
			}
			catch (Exception ex)
			{
				Log.Error(ex);
				throw;
			}     
        }

        static readonly ILog Log = LogManager.GetLogger(typeof(FdkTrade));


        public static double[] GetPositionAgentCommission(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.AgentCommission);
        }
        
        public static double[] GetPositionBuyAmount(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.BuyAmount);
        }
        public static double[] GetPositionBuyPrice(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.BuyPrice ?? double.NaN);
        }
        public static double[] GetPositionCommission(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.Commission);
        }
        
        public static double[] GetPositionProfit(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
			return tradeData.SelectToArray(it => it.Profit?? double.NaN);
        }
        public static double[] GetPositionSellAmount(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.SellAmount);
        }
        public static double[] GetPositionSellPrice(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.SellPrice ?? double.NaN);
        }
        
        public static double[] GetPositionSettlementPrice(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.SettlementPrice);
        }

        
        public static double[] GetPositionSwap(string varName)
        {
            var tradeData = FdkVars.GetValue<Position[]>(varName);
            return tradeData.SelectToArray(it => it.Swap);
        }

		public static string[] GetPositionSymbol(string varName)
		{
			var tradeData = FdkVars.GetValue<Position[]>(varName);
			return tradeData.SelectToArray(it => it.Symbol);
		}
    }
}

File: RPackage/R/FdkPositions.R
#' Gets the account positions
#' 
ttTrade_Position <- function(){
  symInfo = GetTradePositions()
  
  agentComission = GetPositionAgentCommission(symInfo)
  buyAmount = GetPositionBuyAmount(symInfo)
  buyPrice = GetPositionBuyPrice(symInfo)
  comission = GetPositionCommission(symInfo)
  profit = GetPositionProfit(symInfo)
  sellAmount = GetPositionSellAmount(symInfo)
  sellPrice = GetPositionSellPrice(symInfo)
  settlementPrice = GetPositionSettlementPrice(symInfo)
  swap = GetPositionSwap(symInfo)
  symbol = GetPositionSymbol(symInfo)
  
  UnregisterVar(symInfo)
  
  data.table(agentComission, buyAmount, buyPrice, comission, profit,
    sellAmount, sellPrice, settlementPrice, swap, symbol
  )
}
#' Get trade history
GetTradePositions <- function() {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetTradePositions')
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionAgentCommission <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionAgentCommission', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionBuyAmount <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionBuyAmount', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionBuyPrice <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionBuyPrice', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionCommission <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionCommission', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionProfit <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionProfit', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionSellAmount <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionSellAmount', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionSellPrice <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionSellPrice', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionSettlementPrice <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionSettlementPrice', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionSwap <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionSwap', symInfo)
}

#' Get trade comission
#' @param symInfo RHost variable that stores the array
GetPositionSymbol <- function(symInfo) {
  rClr::clrCallStatic('FdkRHost.FdkPosition', 'GetPositionSymbol', symInfo)
}
